crates/biome_js_type_info/src/globals_builder.rs

```diff
@@ -0,0 +1,61 @@
+//! Builder for constructing GlobalsResolver with forward references support.
+
+use std::sync::Arc;
+
+use crate::{NUM_PREDEFINED_TYPES, TypeData, TypeId, TypeStore};
+
+use super::globals::GlobalsResolver;
+
+/// Builder for constructing a GlobalsResolver
+pub struct GlobalsResolverBuilder {
+    /// Types being built. None = reserved but not yet filled.
+    types: Vec<Option<TypeData>>,
+}
+
+impl GlobalsResolverBuilder {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            types: vec![None; capacity],
+        }
+    }
+
+    /// Fill a previously reserved type slot with actual type data.
+    pub fn set_type_data(&mut self, id: TypeId, data: TypeData) {
+        let index = id.index();
+        assert!(
+            index < self.types.len(),
+            "TypeId {index} out of bounds (len: {})",
+            self.types.len()
+        );
+        assert!(
+            self.types[index].is_none(),
+            "Type at index {index} already set"
+        );
+        self.types[index] = Some(data);
+    }
+
+    /// Build the final GlobalsResolver.
+    pub fn build(self) -> GlobalsResolver {
+        let types: Vec<Arc<TypeData>> = self
+            .types
+            .into_iter()
+            .enumerate()
+            .map(|(i, opt)| {
+                Arc::new(opt.unwrap_or_else(|| {
+                    // Make sure we only reserve just enough for what we need
+                    panic!("Type at index {i} was reserved but never filled")
```

Comment:
Take note to remove the panic

Reply:
I defaulted to Unknown. Is that reasonable?

```
        let types: Vec<Arc<TypeData>> = self
            .types
            .into_iter()
            .map(|opt| Arc::new(
                opt.unwrap_or(TypeData::Unknown)))
            .collect();
```